### DISCUSSION TITLE
custom authentication

### DISCUSSION THREAD

**Discussion by pkalvagit** [open-telemetry/opentelemetry-collector#6207] [Q&A]
I am new to opentelemenrty. It may be basic question.. I would like to build a custom collector with custom authentication mechanism and also want to include routing and load balancing features along with exports. I have gone through read me of https://github.com/open-telemetry/opentelemetry-collector/tree/main/cmd/builder and  https://opentelemetry.io/docs/collector/custom-auth/ . Can someone help to step to build custom authentication and custom collector.

**Comment by jpkrohling** [comment]
With which step are you having trouble? You should develop your custom authenticator (server authenticator, in your case), and include it in the "extensions" part of your collector manifest.

**Reply by pkalvagit** [reply]
how to make custom authenticator build part of custom collector in local env? Is there any blog which show step by step approach or if you are able to share also fine..

**Reply by jpkrohling** [reply]
You can prepare a manifest for the builder with a `replace` statement pointing to the local directory containing the module. Your build will then include your custom component.

**Reply by pkalvagit** [reply]
Is it possible to share an example?

**Reply by pkalvagit** [reply]
tried with replace but without my custom autentication also custom collector getting error: Error: failed to update go.mod: exit status 1. Output: "go: finding module for package github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextension\ngo: finding module for package github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter\ngo: found github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter in github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter v0.61.0\ngo: found github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextension in github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextension v0.61.0\ngo.opentelemetry.io/collector/cmd/otelcorecol imports\n\tgo.opentelemetry.io/collector/exporter/otlpexporter imports\n\tgo.opentelemetry.io/collector/config/configgrpc imports\n\tgo.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc tested by\n\tgo.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.test imports\n\tgoogle.golang.org/grpc/interop imports\n\tgolang.org/x/oauth2/google imports\n\tcloud.google.com/go/compute/metadata: ambiguous import: found package cloud.google.com/go/compute/metadata in multiple modules:\n\tcloud.google.com/go v0.65.0 (C:\\Users\\pkalva\\go\\pkg\\mod\\cloud.google.com\\go@v0.65.0\\compute\\metadata)\n\tcloud.google.com/go/compute v1.10.0 (C:\\Users\\pkalva\\go\\pkg\\mod\\cloud.google.com\\go\\compute@v1.10.0\\metadata)\n"

**Reply by pkalvagit** [reply]
I am able resolve above loud.google.com/go/compute/metadata: ambiguous import. I added builder-config as below
`extensions:
  - import: go.opentelemetry.io/collector/extension/ballastextension
    gomod: go.opentelemetry.io/collector v0.61.0
  - import: go.opentelemetry.io/collector/extension/zpagesextension
    gomod: go.opentelemetry.io/collector v0.61.0
  - gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextension v0.61.0"
  #- gomod: "github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckextension v0.59.0"
  #  gomod: go.opentelemetry.io/collector v0.61.0
  - gomod: "testbasicauthentication"
    path: "/opentelemetry-collector-contrib/extension/testbasicauthentication"`

getting Error: failed to update go.mod: exit status 1. Output: "go: errors parsing go.mod:\nC:\\Users\\XYZ\\opentelemetry-examples\\dist\\otelcorecol\\go.mod:11:2: usage: require module/path v1.2.3\nC:\\Users\\XYZ\\opentelemetry-examples\\dist\\otelcorecol\\go.mod:17:2: require github.com/open-telemetry/opentelemetry-collector-contrib/exporter/testbasicauthentication: version \"V0.61.0\" invalid: unknown revision V0.61.0\n"

i tried with import inplace gomod but not working. Any suggestions?
